### PR TITLE
added exists attribute, reflecting whether we have data at a given path

### PR DESCRIPTION
**Pull request by christophe-g**
Replace and close https://github.com/firebase/polymerfire/pull/177. 
Follows #33 discussion. 

An "empty-result" event is fired by firebase-document or firebase-query to notify that the path does not hold any data.

### CODE REVIEW

**Comment by christophe-g** [comment]
Any new on this one ?
Thanks.

**Comment by JosefJezek** [comment]
@mbleigh FYI

**Comment by tjmonsi** [comment]
@JosefJezek can you check it again as it has some conflicts with firebase-query

**Comment by christophe-g** [comment]
@tjmonsi - thanks. 
@mbleigh - for your review ; )

**Comment by christophe-g** [comment]
@tjmonsi  - thanks !
@andrewspy - for info

> Why not use the exist() method to show that a value on the snapshot really exists given a path?

I am not sure which `exist` method you are referring to - if there is one, the reason for not using is certainly that this PR predates an `exist`  implementation. 

If you are referring to `snapshot.exist`, the reason is to better integrate the PR within existing code flow. 

For instance, as we already have this ([with `value == null` which is equivalent to `snapshot.exists()`](https://firebase.google.com/docs/reference/js/firebase.database.DataSnapshot#val) ):
```js
    __onFirebaseValue: function(snapshot) {
      var value = snapshot.val();
      if (value == null) {
        value = this.zeroValue;
        this.__needSetData = true;
      }

```
It is pretty natural to set `exists` to `false` there, without adding one additional check.


> I could edit this PR and add the ready attribute as well in this PR if you want or you can do it. :)

I would rather keep `exists` instead of `ready` as it conveys the idea of whether data exists or not. `ready` just indicates that database round-trip is completed - but I am open to discuss it ; )

**Comment by tjmonsi** [comment]
```
var value = snapshot.val();
      if (value == null) {
        value = this.zeroValue;
        this.__needSetData = true;
      }
```
I think we can change this as 
```
this.exists = snapshot.exist();
if (!this.exists) {
  ...
}
```
I'm still trying to find time for this but I think I can squeeze this in today with `ready`
If that's alright with you, I'll do the changes.

**Comment by andrewspy** [comment]
@christophe-g @tjmonsi In my opinion, both `exists` and `ready` serve a different purpose.

`exists` should refer to the `snapshot.exists()` in official firebase Nodejs API to make the API complete, but in Polymerfire land, it is useful for a developer to know when data is `ready`, with the data guarantee available for data-binding and consumption purpose (assuming the data does exist).

In short, `exists` is meant to use pragmatically, while `ready` is meant to use with Polymer + Polymerfire specifically for data-binding and consumption purpose.

**Comment by christophe-g** [comment]
@tjmonsi 

> If that's alright with you, I'll do the changes.

That is alright - I am pretty stretched!

**Comment by tjmonsi** [comment]
@christophe-g gotyah. Will work on it in a bit

**Comment by merlinnot** [comment]
I see the reasoning behind this, but there are few things I’d like to discuss:

1. Since the value of `exists` is based on a value of `data`, wouldn’t it be better to make it a computed property? Since we cannot use `setProperties` here to ensures the property changes run as a coherent set I think it would be better to set the value of `exists` after changes to the data itself are made.
2. Do we really need this event? Polymer provides us with `on-exists-changed`, isn’t that enough?
3. I’d personally prefer to set this value to undefined if it’s unknown. Why did you decide to go with `null`?

**Comment by christophe-g** [comment]
@merlinnot - thanks for looking at this. 

1. If I remember correctly, I was worried about performance penalties in adding computed properties listening to data changes. As we already indirectly observe data modification in firebase callbacks (`__onFirebaseValue` for `firebase-document` and `__onChildAdded`, `__onChildRemoved` for `firebase-query`), I am assuming (without testing) that performance penalties are minimum with current implementation.

2. I am not using th `empty-result` event. It was suggested in the related #33 discussion. We can remove it as far as I am concerned.

3. Polymer 1.0 does not fire computed/observed properties when one of the parameters are undefined.  For the 1.0 branch - as `exists` property notifies and is potentially bound to external components - setting `undefined` will partially break the purpose of this PR. I am fine with `undefined` for the 2.0 branch.

File: firebase-query.html [hunk 1]
```diff
@@ -311,6 +312,8 @@
               * the callback if the query changes
               */
             query.on('value', this.__onFirebaseValue, this.__onError, this);
+            this._setExists(null);
```
Comment (merlinnot) [inline firebase-query.html]: This one can be moved immediately after `query.off`s

File: firebase-query.html [hunk 2]
```diff
@@ -375,6 +380,10 @@
               this.syncToMemory(function() {
                 this.splice('data', this.__indexFromKey(key), 1);
               });
+              if (this.data.length === 0) {
```
Comment (merlinnot) [inline firebase-query.html]: Hm... what if `data` is changed in some property effect? Don't you think we should get the length of `data` before calling `this.splice`?

**Comment by merlinnot** [comment]
1. I see, that's ok.
2. Let's remove it.
3. Ok, I'm fine with that. I guess we still have to support v1 guys ;)

Could you please add some basic tests? Just a few :smiley:

@tjmonsi what about `ready`? ;)

File: firebase-query.html (hunk 2 above)
Comment (christophe-g) [reply firebase-query.html]: We are just assessing whether we have data or not. Sounds correct to me,

**Comment by christophe-g** [comment]
@tjmonsi @merlinnot - ready for review

**Comment by christophe-g** [comment]
@merlinnot - any news on this ?

> Just a few test :grinning: :smiley:

There are even tests !
Thnks

**Comment by merlinnot** [comment]
@christophe-g I've seen all the work you've done and I'm really happy to see progress on this, but I'm not an owner of this repository and even if I would I'd like to wait for more people to take a look on this. 

Unfortunately, it's kinda hard to reach any of the owners lately, I don't really know why. We have to wait for @tjmonsi or @mbleigh to have some time to take a look at this. I've even tried to tweet the Firebase team but got no response.

**Pull request merged**